vidarr-core/src/main/java/ca/on/oicr/gsi/vidarr/core/ResourceOptimizingPriorityScorer.java

```diff
@@ -0,0 +1,238 @@
+package ca.on.oicr.gsi.vidarr.core;
+
+import ca.on.oicr.gsi.vidarr.PriorityScorer;
+import io.undertow.server.HttpHandler;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+public class ResourceOptimizingPriorityScorer implements PriorityScorer {
+
+  protected record WorkflowRunScore(String workflowName, String workflowVersion, String vidarrId,
+                                    int currentPriority, int originalPriority) implements
+      Comparable<WorkflowRunScore> {
+
+    @Override
+    public int compareTo(WorkflowRunScore workflowRunScore) {
+      int result = Integer.compare(workflowRunScore.currentPriority, this.currentPriority);
+      if (result == 0) {
+        result = this.vidarrId.compareTo(workflowRunScore.vidarrId);
+      }
+      return result;
+    }
+  }
+
+  private final SortedSet<WorkflowRunScore> active = new TreeSet<>();
+  private boolean useCustom;
+  private int globalMaxInFlight, maxInFlightPerWorkflow;
+  // TODO 'mode' setting for workflow vs workflow version?
```

Comment:
Did you end up deciding how to resolve this TODO?

Reply:
No, I wanted reviewers opinion on it

Reply:
I had implemented this as using by-workflow-version as that's what Mike had requested, but looking at the vidarr config, we're only using by-workflow, so I'll change the implementation to that and we can expand later

Reply:
Done